dimscord/dispatch.nim

```diff
@@ -411,31 +410,23 @@ proc messageUpdate(s: Shard, data: JsonNode) {.async.} =
         msg = Message(
             id: data["id"].str,
             channel_id: data["channel_id"].str)
-        oldMessage: Option[Message]
-        exists = false
-
-    if msg.channel_id in s.cache.guildChannels:
-        let chan = s.cache.guildChannels[msg.channel_id]
-
-        if msg.id in chan.messages:
-            msg = chan.messages[msg.id]
-            oldMessage = some move chan.messages[msg.id]
-            exists = true
-
-        msg = msg.updateMessage(data)
-        if msg.id in chan.messages: chan.messages[msg.id] = msg
-    elif msg.channel_id in s.cache.dmChannels:
-        let chan = s.cache.dmChannels[msg.channel_id]
-
-        if msg.id in chan.messages:
-            msg = chan.messages[msg.id]
-            oldMessage = some move chan.messages[msg.id]
-            exists = true
-
-        msg = msg.updateMessage(data)
-        if msg.id in chan.messages: chan.messages[msg.id] = msg
-
-    s.checkAndCall(MessageUpdate, msg, oldMessage, exists)
+    # Get the messages table if the channel is in the cache
+    var messages {.cursor.} =
+        if msg.channel_id in s.cache.guildChannels:
+            s.cache.guildChannels[msg.channel_id].messages
+        elif msg.channel_id in s.cache.dmChannels:
+            s.cache.dmChannels[msg.channel_id].messages
+        else: initTable[string, Message](0)
+    let
+        exists = msg.id in messages
+        # Use the version from cache if we can
+        oldMsg = if exists: messages[msg.id] else: nil
+        # Update the message from cache, or just use what we have
+        newMsg = if exists: oldMsg.updateMessage(data) else: msg
+    # Update the messages
+    if exists:
+        messages[msg.id] = newMsg
+    s.checkAndCall(MessageUpdate, newMsg, option(oldMsg), exists)
```

Comment:
Should `newMsg` use the full data? I kept it like this since the old code didn't update the message object if nothing existed in cache but should it instead also try to add what data it can to the new message?

Reply:
It should always contain the full data just like the old one. if old message doesnt exist in cache then the oldMsg param should be `none` as always. The new message should have the data equivalent to what was in the raw json data, which I seem to have made a mistake on and rereading the old code it only contains id and channel_id if it doesnt exist in cache. It should be `msg = newMessage(data)`.

I'm assuming that on the discord docs it used to mention that message data in the event did not contain the whole message data before, so hence why it might have led to that line of code. [But seems now currently that's no longer the case after checking the docs.](https://discord.com/developers/docs/topics/gateway-events#message-update)

Reply:
Nice, that makes it a bit simplier. Now it creates a new message and updates the cache